// ts-progress/progress.js
"use strict";
var numeral = require('numeral');
var charm = require('charm')();
charm.pipe(process.stdout);
var Progress = (function () {
    function Progress(_total, pattern, _textColor, _title, _updateFrequency) {
        var _this = this;
        if (_updateFrequency === void 0) { _updateFrequency = 0; }
        this._total = _total;
        this._textColor = _textColor;
        this._title = _title;
        this._updateFrequency = _updateFrequency;
        this._barSize = 20;
        this._percent = 0;
        this._percentIncrease = 0;
        this._current = 0;
        this._start = 0;
        this._elapsed = 0;
        this._remaining = 0;
        this._now = 0;
        this._cycle = 0;
        this._pattern = 'Progress: {bar} | Elapsed: {elapsed} | {percent}';
        this._regex = /(.*?){(.*?)}/g;
        this.start = function () {
            _this._start = new Date().getTime();
            _this._now = new Date().getTime();
            _this._cycle = _this._start;
            _this.renderTitle();
            _this.write();
        };
        this.stop = function () {
            _this._current++;
            _this._percent = 100;
            _this._elapsed = (_this._now - _this._start) / 1000;
            _this._remaining = 0;
            _this.write();
            charm.write("\n");
        };
        this.write = function () {
            charm.erase('line').write("\r");
            var match;
            while (match = _this._regex.exec(_this._pattern)) {
                _this.renderText(match[1]);
                if (match[2].indexOf('.') == -1) {
                    _this.renderPattern(match[2], match[2]);
                }
                else {
                    var tokens = match[2].split('.');
                    if (tokens.length == 4 && tokens[0] === 'bar') {
                        _this.renderBar(tokens[1], tokens[2], tokens[3]);
                    }
                    else if (tokens.length == 2) {
                        _this.renderPattern(match[2], tokens[0], tokens[1]);
                    }
                }
            }
        };
        this.renderElapsed = function (color) {
            _this.renderItem(numeral(_this._elapsed).format('0.0') + 's', color);
        };
        this.renderRemaining = function (color) {
            _this.renderItem(numeral(_this._remaining).format('0.0') + 's', color);
        };
        this.renderMemory = function (color) {
            _this.renderItem(numeral(process.memoryUsage().rss / 1024 / 1024).format('0.0') + 'M', color);
        };
        this.renderPercent = function (color) {
            _this.renderItem(numeral(_this._percent).format('0') + "%", color);
        };
        this.renderCurrent = function (color) {
            _this.renderItem(_this._current.toString(), color);
        };
        this.renderTotal = function (color) {
            _this.renderItem(_this._total.toString(), color);
        };
        this.renderBar = function (colorRemaining, colorDone, size) {
            if (colorRemaining === void 0) { colorRemaining = 'white'; }
            if (colorDone === void 0) { colorDone = 'green'; }
            if (size && size !== _this._barSize)
                _this._barSize = size;
            charm.foreground(colorDone).background(colorDone);
            var done = Math.ceil(((_this._current / _this._total) * _this._barSize));
            charm.write(_this._padding.substr(0, done));
            charm.foreground(colorRemaining).background(colorRemaining);
            charm.write(_this._padding.substr(0, _this._barSize - done));
            charm.display('reset');
        };
        this._patternMapping = {
            'bar': this.renderBar,
            'elapsed': this.renderElapsed,
            'remaining': this.renderRemaining,
            'memory': this.renderMemory,
            'percent': this.renderPercent,
            'current': this.renderCurrent,
            'total': this.renderTotal,
        };
        this.renderPattern = function (pattern, item, color) {
            var renderer = _this._patternMapping[item];
            if (renderer) {
                renderer(color);
            }
            else {
                charm.write(pattern);
            }
        };
        this.renderItem = function (item, color) {
            if (color)
                charm.foreground(color).write(item).display('reset');
            else
                charm.write(item);
        };
        this.renderText = function (text) {
            if (_this._textColor)
                charm.display('bright').foreground(_this._textColor).write(text).display('reset');
            else
                charm.display('bright').write(text).display('reset');
        };
        this.renderTitle = function () {
            if (_this._title && _this._title !== '') {
                charm.display('bright').write(_this._title).display('reset');
                charm.write("\n");
            }
        };
        this._padding = new Array(300).join(' ');
        if (pattern)
            this._pattern = pattern;
        //this._padding = new Array(300).join('▒');
        this._percentIncrease = 100 / _total;
        this.start();
    }
    Progress.prototype.update = function () {
        this._now = new Date().getTime();
        if (this._current === this._total - 1) {
            this.stop();
        }
        else {
            this._current++;
            this._percent += this._percentIncrease;
            if (!this.skipStep()) {
                this._elapsed = (this._now - this._start) / 1000;
                this._remaining = (this._elapsed / this._current * (this._total - this._current));
                this.write();
            }
        }
    };
    Progress.prototype.skipStep = function () {
        if (this._updateFrequency == 0)
            return false;
        var elapsed = this._now - this._cycle;
        if (elapsed < this._updateFrequency) {
            return true;
        }
        else {
            this._cycle = this._now;
            return false;
        }
    };
    return Progress;
}());
exports.Progress = Progress;
module.exports = Progress;
